src/replication.c

```diff
@@ -2819,8 +3599,32 @@ void syncWithPrimary(connection *conn) {
         server.repl_transfer_fd = dfd;
     }
 
+    /* Using rdb-connection sync, the primary responded +FULLSYNCNEEDED. We need to 
+     * initialize the RDB connection. */
+    if (psync_result == PSYNC_FULLRESYNC_RDB_CONN) {
+        /* Create a full sync connection */
+        server.repl_rdb_transfer_s = connCreate(connTypeOfReplication());
+        if (connConnect(server.repl_rdb_transfer_s, server.primary_host, server.primary_port,
+                    server.bind_source_addr, fullSyncWithPrimary) == C_ERR) {
+            serverLog(LL_WARNING,"Unable to connect to Primary: %s",
+                    connGetLastError(server.repl_transfer_s));
+            connClose(server.repl_rdb_transfer_s);
+            server.repl_rdb_transfer_s = NULL;
+            goto error;
+        }
+        if (connSetReadHandler(conn, NULL) == C_ERR) {
+            char conninfo[CONN_INFO_LEN];
+            serverLog(LL_WARNING,
+                "Can't clear main connection handler: %s (%s)",
+                strerror(errno), connGetInfo(conn, conninfo, sizeof(conninfo)));
+            goto error;
+        }
+        server.repl_rdb_conn_state = REPL_RDB_CONN_SEND_HANDSHAKE;
+        return;
+    }
     /* Setup the non blocking download of the bulk file. */
-    if (connSetReadHandler(conn, readSyncBulkPayload) == C_ERR) {
+    else if (connSetReadHandler(conn, readSyncBulkPayload) == C_ERR)
```

Comment:
This `else` is a bit confusing IMO.
```suggestion
    if (connSetReadHandler(conn, readSyncBulkPayload) == C_ERR)
```